tutorials/navigation/navigation_using_navigationmeshes.rst

```diff
@@ -229,95 +229,152 @@ There are some common user problems and important caveats to consider when creat
             source_geometry,
-            root_node
+            root_node,
+            callback_parsing
         )
 
-        # Bake the navigation mesh with the source geometry data.
+    func on_parsing_done() -> void:
+        # If we did not parse a TileMap with navigation mesh cells we may now only
+        # have obstruction outlines so add at least one traversable outline
+        # so the obstructions outlines have something to "cut" into.
+        source_geometry.add_traversable_outline(PackedVector2Array([
+            Vector2(0.0, 0.0),
+            Vector2(500.0, 0.0),
+            Vector2(500.0, 500.0),
+            Vector2(0.0, 500.0)
+        ]))
+
+        # Bake the navigation mesh on a thread with the source geometry data.
         NavigationServer2D.bake_from_source_geometry_data_async(
             navigation_mesh,
-            source_geometry
+            source_geometry,
+            callback_baking
         )
 
-        # Create a new navigation region and update the region with prepared navigation mesh.
-        var region_rid: RID = NavigationServer2D.region_create()
+    func on_baking_done() -> void:
+        # Update the region with the updated navigation mesh.
         NavigationServer2D.region_set_navigation_polygon(region_rid, navigation_mesh)
-        NavigationServer2D.region_set_enabled(region_rid, true)
-        NavigationServer2D.region_set_map(region_rid, get_world_2d().get_navigation_map())
 
  .. code-tab:: gdscript 3D GDScript
 
     extends Node3D
 
-    func parse_and_bake_navigation_mesh_and_create_region() -> void:
-        # Note, the navigation mesh is not baked here, it only holds the parse parameters.
-        var navigation_mesh: NavigationMesh = NavigationMesh.new()
-        # Create the source geometry that will hold the parsed data.
-        var source_geometry := NavigationMeshSourceGeometryData3D.new()
-        # The Node where the parsing should start traversing the SceneTree.
+    var navigation_mesh: NavigationMesh
+    var source_geometry : NavigationMeshSourceGeometryData3D
+    var callback_parsing : Callable
+    var callback_baking : Callable
+    var region_rid: RID
+
+    func _ready() -> void:
+        navigation_mesh = NavigationMesh.new()
+        navigation_mesh.agent_radius = 0.5
+        source_geometry = NavigationMeshSourceGeometryData3D.new()
+        callback_parsing = on_parsing_done
+        callback_baking = on_baking_done
+        region_rid = NavigationServer3D.region_create()
+
+        # Enable the region and set it to the default navigation map.
+        NavigationServer3D.region_set_enabled(region_rid, true)
+        NavigationServer3D.region_set_map(region_rid, get_world_3d().get_navigation_map())
+
+        # Some mega-nodes like GridMap are often not ready on the first frame.
+        # Also the parsing needs to happen on the main-thread.
+        # So do a deferred call to avoid common parsing issues.
+        parse_source_geometry.call_deferred()
```

Comment:
This, specifically. Again, better to document this than not, but maybe something we should file somewhere to see if we can improve UX here.